### PR TITLE
JIT: Enhance metrics reported by superpmi diffs

### LINKED ISSUE #70350: SPMI CI runs need to surface context misses
If there are a significant number of misses in SPMI runs then the results we show on the "Extensions" page are very misleading, both for the ASM diffs and TP impact. We need to surface the misses so that we can more easily evaluate whether a normal jit-diff run is needed.

### PR DESCRIPTION
**Pull request by jakobbotsch**
* Report the total number of contexts, minopts contexts and fullopts
  contexted processed
* Report number of successful and missing contexts
* Report asmdiffs and tpdiffs for minopts/fullopts separately

Fixes https://github.com/dotnet/runtime/issues/70350
Contributes to https://github.com/dotnet/runtime/issues/73506

### CODE REVIEW

**Comment by jakobbotsch** [comment]
@dotnet/jit-contrib I'm decently happy with what I have now. It looks like the following for a change with diffs: https://dev.azure.com/dnceng-public/public/_build/results?buildId=14828&view=ms.vss-build-web.run-extensions-tab

Currently running it to make sure it does not look horrible when there are no diffs, but otherwise this is ready for review. @kunalspathak can you take a look?

**Review by kunalspathak — COMMENTED** [review]
More details are always good to have, but I honestly think that the extension page is getting very crowded, and it is hard to follow the most common information that we want to see. 

I think the confusing part is when you extend "Details" and at one point I lose track of whose details I was looking for.

File: src/coreclr/scripts/superpmi_diffs_summarize.py [hunk 1]
```diff
@@ -89,39 +89,8 @@ def append_diff_file(f, arch, file_name, full_file_path, asmdiffs):
         else:
             diffs_found = True
             f.write("## {} {}\n".format(diff_os, diff_arch))
-
```
[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/scripts/superpmi_diffs_summarize.py:92]: The sections are printed by superpmi.py now.

File: src/coreclr/scripts/superpmi.py [hunk 2]
```diff
@@ -1816,21 +1824,100 @@ async def create_one_artifact(jit_path: str, location: str, flags) -> str:
...
+
+                if any(has_diff for (_, _, _, has_diff, _) in asm_diffs):
+                    def write_pivot_section(col, open):
+                        write_fh.write("\n<details{}>\n".format(" open" if open else ""))
+                        sum_base = sum(int(base_metrics[col]["Diffed code bytes"]) for (_, base_metrics, _, _, _) in asm_diffs)
+                        sum_diff = sum(int(diff_metrics[col]["Diffed code bytes"]) for (_, _, diff_metrics, _, _) in asm_diffs)
+
+                        write_fh.write("<summary>{} ({} bytes)</summary>\n\n".format(col, format_delta(sum_base, sum_diff)))
+                        write_fh.write("|Collection|Base size (bytes)|Diff size (bytes)|\n")
+                        write_fh.write("|---|--:|--:|\n")
```
[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/scripts/superpmi.py:1861]: I'm not sure if this is possible with markdown tables in Azure Pipelines. I agree the default style there is not great, though with only three columns I haven't really had this particular problem.
Maybe switching to HTML tables would work.

File: src/coreclr/tools/superpmi/superpmi/jitinstance.cpp [hunk 3]
```diff
@@ -335,6 +339,14 @@ JitInstance::Result JitInstance::CompileMethod(MethodContext* MethodToCompile, i
         CORINFO_OS os             = CORINFO_WINNT;
 
         pParam->pThis->mc->repCompileMethod(&pParam->info, &pParam->flags, &os);
+        CORJIT_FLAGS jitFlags;
+        pParam->pThis->mc->repGetJitFlags(&jitFlags, sizeof(jitFlags));
+
+        *pParam->isMinOpts =
+            jitFlags.IsSet(CORJIT_FLAGS::CORJIT_FLAG_DEBUG_CODE) ||
+            jitFlags.IsSet(CORJIT_FLAGS::CORJIT_FLAG_MIN_OPT) ||
+            jitFlags.IsSet(CORJIT_FLAGS::CORJIT_FLAG_TIER0);
```
[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/tools/superpmi/superpmi/jitinstance.cpp:348]: Today they are practically the same, I think. They both lead to `OptimizarionDisabled()` returning true.

**Comment by jakobbotsch** [comment]
> More details are always good to have, but I honestly think that the extension page is getting very crowded, and it is hard to follow the most common information that we want to see.

What is the most common information you want to see? Currently it is showing the overall diffs for each collection by default, which is what I want to see, but required multiple clicks before to individually look at.

I do agree it is still a bit overwhelming to look at, but I'm not sure how to present it better. Part of this is that some of this information is new but is information we really ought to always have looked at as part of evaluating changes (in particular the MISSED contexts).

File: src/coreclr/tools/superpmi/superpmi/jitinstance.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jakobbotsch) [reply src/coreclr/tools/superpmi/superpmi/jitinstance.cpp:348]: (Also note that `Compiler::Options::MinOpts` returns true in tier-0 compilations, so it's not wrong to say that tier-0 compilations are MinOpts)

**Comment by jakobbotsch** [comment]
> the extension page is getting very crowded

Do you think it would be better if I leave the "Overall" sections unexpanded by default? Then you at least aren't presented with so many tables immediately when you go to look.

File: src/coreclr/tools/superpmi/superpmi/jitinstance.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (kunalspathak) [reply src/coreclr/tools/superpmi/superpmi/jitinstance.cpp:348]: I see, I was checking yesterday and comparing it with "debuggable code" which is different. I was thinking that "debuggable code" is min-opts.

**Comment by kunalspathak** [comment]
> Do you think it would be better if I leave the "Overall" sections unexpanded by default? Then you at least aren't presented with so many tables immediately when you go to look.

Yes, that's what I mean by important information shown by default and everything else is hidden.

**Comment by jakobbotsch** [comment]
> Yes, that's what I mean by important information shown by default and everything else is hidden.

Well, that's already how it was. It showed you the diffs by default and hid everything else.

The new one that hides the diffs by default looks like this: https://dev.azure.com/dnceng-public/public/_build/results?buildId=16291&view=ms.vss-build-web.run-extensions-tab

Since we show the sum of diffs now it seems fine to me to hide the per-collection diffs. What do you think?

**Comment by kunalspathak** [comment]
1. I mean, here I am not sure showing `> MinOpts (+0 bytes)`, `> FullOpts (+194,496 bytes)` by default is of any value. It is good information to have, but not be default IMO. We can just stick with `Overall` numbers.
image

If someone sees those numbers, they anyway don't have clue when they expand "Details diff" to see where they are coming from? I would suggest putting it few clicks under should be fine. For instance, we might not have cases where MinOpts code size diffs are changing frequently, but perhaps they might show up once in a while on TP metrics. 

2.  > The "sum of diffs" is not really a meaningful number to base anything on, but it does help to guide where one should drill in.

image
How about removing the "Base size" column? What are we going to do knowing that? Instead `Diff size (bytes)` and `Diff %` might be better?

3.  I always felt spacing between `Details` and next configuration say `OSX arm64` show be more to distinguish that we are starting the next section. Without that it looks more congested when expanded. 

image

Probably a fancy line separator might help.

If we are ever able to post back these results to PR as comments, I would like to include some asmdiff examples of each category too and then, we will have to find some more space to have them. 

Again, these are just some suggestions, and it might take multiple rounds to make it to the point where it proves readable for everyone. Feel free to ignore some of the suggestions.

**Comment by jakobbotsch** [comment]
> 1. I mean, here I am not sure showing `> MinOpts (+0 bytes)`, `> FullOpts (+194,496 bytes)` by default is of any value. It is good information to have, but not be default IMO. We can just stick with `Overall` numbers.

I strongly disagree. Having the diffs separated out into these categories is quite important to me. Code size regressions in MinOpts are not as important as code size regressions in FullOpts. TP regressions in MinOpts are more important than TP regressions in FullOpts. If I'm making a transformation that is enabled only when optimizing then "Overall" is a misleading category -- I am interesting in its impact only on optimized contexts. I am also interested in verifying that there is no impact on MinOpts.

I'd be fine with hiding it under "Details", but I think the separated categories are quite important. What does the rest of @dotnet/jit-contrib think? Should they be shown by default or hidden away?

> If someone sees those numbers, they anyway don't have clue when they expand "Details diff" to see where they are coming from?

We can fix jit-analyze separately. At least this makes it clear that there is impact on these categories of contexts and that further investigation is warranted.

> How about removing the "Base size" column? What are we going to do knowing that? Instead `Diff size (bytes)` and `Diff %` might be better?

This was how I first had it, but I quickly saw that "Diff %" is almost always too small to give any meaningful information. For most changes I looked at it is 0.00%. This very large forward sub change is in the 0.01% - 0.02% range. So this was an alternative way to give some notion of relative impact. Not sure if there's a better way (permille?) -- I personally have internalized what a "large diff" is based on just the absolute diff after working on the JIT for a while.

> I always felt spacing between Details and next configuration say OSX arm64 show be more to distinguish that we are starting the next section. Without that it looks more congested when expanded.

I agree, let me try to add some more separation.

**Comment by SingleAccretion** [comment]
> Should they be shown by default or hidden away?

I agree with Jakob that the minopts/fullopts separation should be shown prominently.

I also agree it's a bit unfortunate getting to the jit-analyze output is not so quick, since it contain the outliers (large relative regressions), but I don't think it's a big deal.

Suggestion: grey out (don't color) `0.00%` for TP diffs - green `-0.00%` look a bit strange.

**Comment by kunalspathak** [comment]
> Code size regressions in MinOpts are not as important as code size regressions in FullOpts. TP regressions in MinOpts are more important than TP regressions in FullOpts. 

That's exactly what I stated: "For instance, we might not have cases where MinOpts code size diffs are changing frequently, but perhaps they might show up once in a while on TP metrics." and if we agree that why not hide `code size diffs` of MinOpts and show `TP diff` instead by default?

I think I am trying to convey that we should just surface minimum information on top that matters 80% of time and rest all can be hidden behind few clicks.

**Comment by jakobbotsch** [comment]
> if we agree that why not hide `code size diffs` of MinOpts and show `TP diff` instead by default?

So is your suggestion to move the MinOpts section under Details for asmdiffs, and the FullOpts section under Details for the TP diffs? But retain Overall/FullOpts and OverAll/MinOpts respectively? It's odd to me to have these sections appear in different places.

FWIW, I agree the diffs themselves are less important if they are expected. But easily being able to see whether there are any unexpected diffs is very important too. I dislike having to go expand various subsections for something that I expect I always want to check.

**Comment by kunalspathak** [comment]
Sounds ok to me. May be adding the extra spacing around (as I suggested above) might make it better. And of course, we can re-iterate with better ideas.

**Comment by AndyAyersMS** [comment]
> we can re-iterate with better ideas.

Right, this is a clear improvement ove what we have now, and once we have some mileage on it we can revise.

My only suggestion would be to try and also record the net +/- size contributions, so it is easier to tell if something is a pure improvement or regression or mixed bag, eg
```
Overall (+190,456 bytes) (+250,000/-59,544)
```

But this can wait. Maybe even in the fullness of time, some sort of empirical PDF/histogram for the gains and losses. I am usually interested in knowing when there are outliers and such... many of my recent changes have a lot of small wins and a few big regressions, for example.

**Review by kunalspathak — APPROVED** [review]

**Comment by jakobbotsch** [comment]
This is how it looks with no diffs: https://dev.azure.com/dnceng-public/public/_build/results?buildId=21852&view=ms.vss-build-web.run-extensions-tab
Unfortunately we have spurious diffs in the coreclr_tests.run collection for arm platforms right now (see https://github.com/dotnet/runtime/pull/75672#issuecomment-1250835017)). But you can see how it should look with no diffs for the Linux/Windows x64 sections.

I think I'll remove the "missed contexts" line when there are no misses, though it's probably a quite rare situation.
